go/p2p/peermgmt.go

```diff
@@ -3,423 +3,1240 @@ package p2p
 
-func (p *p2pPeer) connectWorker(mgr *PeerManager, peerID core.PeerID) {
-	defer func() {
-		close(p.doneCh)
-	}()
-
-	ai, err := nodeToAddrInfo(p.node)
+func (m *PeerManager) handleNode(n *node.Node) {
+	info, err := p2pInfoToAddrInfo(&n.P2P)
 	if err != nil {
-		mgr.logger.Error("failed to get node addresses, not retrying",
+		m.logger.Error("failed to convert node to node info",
 			"err", err,
-			"node_id", p.node.ID,
+			"node_id", n.ID,
 		)
 		return
 	}
 
-	mgr.logger.Debug("updating libp2p gossipsub peer",
-		"node_id", p.node.ID,
-		"peer_id", peerID,
-		"addrs", ai.Addrs,
-	)
+	if info.ID == m.host.ID() {
+		return
+	}
 
-	bctx := backoff.WithContext(cmnBackoff.NewExponentialBackOff(), p.ctx)
-
-	err = backoff.Retry(func() (retError error) {
-		// This is blocking, which is stupid.
-		if perr := mgr.host.Connect(p.ctx, *ai); perr != nil {
-			switch perr {
-			case context.Canceled, context.DeadlineExceeded:
-				return backoff.Permanent(perr)
-			default:
-				// This could check if error is a non-temporary
-				// `net.Error`, but continuing to retry as long
-				// as the host claims that it is available
-				// at that address, is probably ok?
-				mgr.logger.Debug("failed to connect to peer",
-					"err", err,
-					"node_id", p.node.ID,
-					"peer_id", peerID,
-					"addrs", ai.Addrs,
-				)
-			}
-			return perr
+	m.host.Peerstore().AddAddrs(info.ID, info.Addrs, peerstore.AddressTTL)
+
+	defer m.initOnce.Do(func() {
+		close(m.initCh)
+	})
+
+	// Protocol versions cannot be found in the registry so we assume all peers speak the latest
+	// protocols. The peer manager should handle this problem and block or negatively grade peers
+	// that are not up-to-date.
+	if n.HasRoles(node.RoleKeyManager) {
+		for _, rt := range n.Runtimes {
+			pid := rpc.NewRuntimeProtocolID(rt.ID, kmp2p.KeyManagerProtocolID, kmp2p.KeyManagerProtocolVersion)
+			m.addPeerToProtocol(info.ID, pid)
 		}
+	}
 
-		mgr.logger.Debug("connected to peer",
-			"node_id", p.node.ID,
-			"peer_id", peerID,
-		)
+	if n.HasRoles(node.RoleComputeWorker) {
+		for _, rt := range n.Runtimes {
+			pid := rpc.NewRuntimeProtocolID(rt.ID, txsync.TxSyncProtocolID, txsync.TxSyncProtocolVersion)
+			m.addPeerToProtocol(info.ID, pid)
 
-		return nil
-	}, bctx)
+			tid := NewTopicIDForRuntime(m.chainContext, rt.ID, api.TopicKindTx)
+			m.addPeerToTopic(info.ID, tid)
 
-	if err != nil {
-		mgr.logger.Warn("failed to connect to peer, not retrying",
-			"err", err,
-			"node_id", p.node.ID,
-		)
+			tid = NewTopicIDForRuntime(m.chainContext, rt.ID, api.TopicKindCommittee)
+			m.addPeerToTopic(info.ID, tid)
+		}
 	}
+
+	if n.HasRoles(node.RoleStorageRPC) {
+		for _, rt := range n.Runtimes {
+			pid := rpc.NewRuntimeProtocolID(rt.ID, pub.StoragePubProtocolID, pub.StoragePubProtocolVersion)
+			m.addPeerToProtocol(info.ID, pid)
+		}
+		for _, rt := range n.Runtimes {
+			pid := rpc.NewRuntimeProtocolID(rt.ID, storageSync.StorageSyncProtocolID, storageSync.StorageSyncProtocolVersion)
```

Comment:
This should be moved up to compute nodes (all compute nodes are also storage nodes).